core/src/main/scala/akka/persistence/cassandra/query/AllEvents.scala

```diff
@@ -0,0 +1,369 @@
+  final case class UpdatePersistenceIdOffset(pid: PersistenceId, seqNr: SeqNr, directEvents: Boolean)
+
+  private object PidState {
+    val empty: PidState = PidState(0L, Vector.empty, inFlight = false)
+  }
+  private case class PidState(seqNr: SeqNr, bufferedDirect: Vector[PersistentRepr], inFlight: Boolean)
+
+  def props(offset: AllEventsOffset, matchingPersistenceId: PersistenceIdsState.PersistenceId => Boolean): Props =
+    Props(new PersistenceIdsState(offset, matchingPersistenceId))
+
+}
+
+class PersistenceIdsState(offset: AllEventsOffset, matchingPersistenceId: PersistenceIdsState.PersistenceId => Boolean)
+    extends Actor
+    with ActorLogging {
+  import PersistenceIdsState._
+
+  private var pidState: Map[PersistenceId, PidState] = offset.offsets.iterator.map {
+    case (pid, seqNr) => pid -> PidState(seqNr, Vector.empty, inFlight = false)
+  }.toMap
+  // FIXME more efficient Queue?
+  private var pending: Vector[PersistenceId] = pidState.keysIterator.toVector.sorted
+  private var directPending: Vector[PersistenceId] = Vector.empty
+
+  private var throttleDelay: FiniteDuration = Duration.Zero
+  private var queryCount = 0
+  private var updateCount = 0
+  private var resetPendingTime = System.nanoTime()
+
+  log.debug("Starting PersistenceIdsState [{}] with [{}] pids in offset.", self.path.name, offset.offsets.size)
+
+  private def scheduler = context.system.scheduler
+
+  private def nextDirectPending(): Option[(PersistenceId, PidState)] = {
+    if (directPending.isEmpty) {
+      None
+    } else {
+
+      val iter = directPending.iterator
+      @tailrec def findNext(): Option[(PersistenceId, PidState)] = {
+        if (iter.isEmpty) {
+          None
+        } else {
+          val pid = iter.next()
+          val state = pidState(pid)
+          if (state.inFlight || state.bufferedDirect.isEmpty) {
+            findNext()
+          } else {
+            // FIXME filterNot is not very efficient, better mutable collection for this purpose?
+            directPending = directPending.filterNot(_ == pid)
+            Some(pid -> state)
+          }
+        }
+      }
+
+      findNext()
+    }
+  }
+
+  private def nextPending(): Option[(PersistenceId, PidState)] = {
+    if (pending.isEmpty && pidState.nonEmpty) {
+      val durationMs = (System.nanoTime() - resetPendingTime).nanos.toMillis
+      log.info(
+        "Reset pending pids, size [{}]. Round took [{} ms]. Found events for [{}] pids.",
+        pidState.size,
+        durationMs,
+        updateCount)
+      resetPendingTime = System.nanoTime()
+      if (updateCount == 0) {
+        // delay every 100th query
+        // measured 10000 pids: 6 seconds without throttling, 1 minute with this (500 ms / 100) throttling
+        // measured 100000 pids: 70 seconds without throttling, 10 minutes with this (500 ms / 100) throttling
+        throttleDelay = 500.millis // FIXME config
```

Comment:
I'm worried that it would generate too many queries in idle state. Maybe we could have another mechanism to publish LRU pid->seqNr so that the scanning can be even more infrequent. DData with some data type for LRU pid->seqNr comes to mind.